brickstrap.sh

```diff
@@ -349,70 +429,124 @@ function configure-packages () {
-function create-report() {
-    DUMP_INFO_HOOK_SCRIPT="${BOARDDIR}/custom-report.sh"
-    if [ -r "${DUMP_INFO_HOOK_SCRIPT}" ]; then
-        info "Running custom reporting script..."
-        . "${DUMP_INFO_HOOK_SCRIPT}"
-        info "Done with custom reporting script."
+function brp_run-hooks() {
+    if [ ! -d "${ROOTDIR}" ]; then
+        fail "${ROOTDIR} does not exist."
+    elif br_list_directories "hooks" >/dev/null; then
+        info "Running hooks..."
+        br_for_each_path_iterate_directories "$(br_list_directories "hooks")" \
+            brp_run_hook_impl
     else
-        info "Skipping custom report, no such script. (${DUMP_INFO_HOOK_SCRIPT})"
+        info "Skipping hooks, no such directory: hooks"
     fi
 }
 
-function create-tar() {
+# Runs a status/config info reporting hook, to be called at the end of the
+# brickstrap process. This permits the user to aggregate important info about
+# the build in a single, convenient report. (E.g. root passwd, default account
+# username+password, hostname, key fingerprints?)
+function brp_create-report() {
+    if br_list_paths "custom-report.sh" -r >/dev/null; then
+        info "Running custom reporting scripts..."
+        br_for_each_path "$(br_list_paths custom-report.sh -r)" \
+            brp_run_hook_impl 'executing'
+        info "Done with custom reporting scripts."
+    else
+        info "Skipping custom report, no such scripts: custom-report.sh"
+    fi
+}
+
+function brp_copy_to_tar_only() {
+    cp -r "$1/." "${ROOTDIR}/tar-only/"
+}
+
+function brp_create-tar() {
     info "Creating tar of rootfs"
     debug "ROOTDIR: ${ROOTDIR}"
     debug "TARBALL: ${TARBALL}"
     [ ! -d "${ROOTDIR}" ] && fail "${ROOTDIR} does not exist."
     [ -z ${FORCE} ] && [ -f "${TARBALL}" ] \
 	    && fail "${TARBALL} exists. Use -f option to overwrite."
-    # need to generate tar inside fakechroot so that absolute symlinks are correct
     info "creating tarball ${TARBALL}"
-    EXCLUDE_LIST=/host-rootfs/${BOARDDIR}/tar-exclude
-    info "Excluding files:
-$(${CHROOTCMD} cat ${EXCLUDE_LIST})"
+
+    info "Excluding files: "
+    BRP_EXCLUDE_LIST="${ROOTDIR}/tar-exclude"
+    br_cat_files tar-exclude | tee "$BRP_EXCLUDE_LIST"
+
+    # need to generate tar inside fakechroot
+    # so that absolute symlinks are correct
+
     ${CHROOTCMD} tar cpf /host-rootfs/${TARBALL} \
-        --exclude=host-rootfs --exclude=usr/bin/$(basename ${QEMU_STATIC}) --exclude=tar-only \
-        --exclude-from=${EXCLUDE_LIST} .
-    if [ -d "${BOARDDIR}/tar-only" ]; then
-      cp -r "${BOARDDIR}/tar-only/." "${ROOTDIR}/tar-only/"
-    fi
-    if [ -d "${ROOTDIR}/tar-only" ]; then
-      info "Adding tar-only files:"
-      ${CHROOTCMD} tar rvpf /host-rootfs/${TARBALL} -C /tar-only .
+        --exclude=host-rootfs --exclude=usr/bin/$(basename ${QEMU_STATIC}) \
+        --exclude=tar-only --exclude=${BRP_EXCLUDE_LIST##$ROOTDIR/} \
+        --exclude-from=${BRP_EXCLUDE_LIST##$ROOTDIR} .
```

Comment:
> This behaviour is not new either.

I should know better, I'm the one that put it there. Thanks for reminding me how it works.



> What would be a useful for a follow-up PR is to nest things like tar-only, tar-exclude and host-rootfs underneath a brickstrap subdirectory in $ROOTDIR.

:+1: